### DISCUSSION TITLE
SunbirdED -Portal :: Copy Question Set feature

### DISCUSSION THREAD

**Discussion by princegupta1131** [Sunbird-inQuiry/Community#80] [General Discussions]
in  SunbirdED -Portal on a copy of the Question set, getting 400 bad server error

As per the suggestion, we have done the code level changes to copy the question set.

**1. URL: {{host}}/questionset/v1/copy/{{questionset_id}}
2. Request Body: please use questionset instead of content**

**Post integration api throwing  401 Unauthorised  Error**

**CURL**

`curl 'http://localhost:3000/content/questionset/v1/copy/do_2133979147196088321344' \
  -H 'Accept: application/json' \
  -H 'Accept-Language: en-US,en;q=0.9' \
  -H 'Cache-Control: no-cache' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Cookie: connect.sid=s%3AStewujMBObqoHrK7tIuqqI9m9bi3H49M.vsS4V%2F%2BDqhE5QyAqmaMFT9BmqecCG7zR81kzHORxMew' \
  -H 'Origin: http://localhost:3000' \
  -H 'Pragma: no-cache' \
  -H 'Referer: http://localhost:3000/resources/play/questionset/do_2133979147196088321344' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36' \
  -H 'X-App-Id: local.sunbird.portal' \
  -H 'X-App-Version: 6.0.0' \
  -H 'X-Channel-Id: 01269878797503692810' \
  -H 'X-Device-ID: d22678ad8b5cc0f76f63dc0e7f4d1480' \
  -H 'X-Org-code: 01269878797503692810' \
  -H 'X-Request-ID: 0892cc42-38c3-4214-8d29-c65a20d86191' \
  -H 'X-Session-ID: RB7pvdtVRopF24OkoGYEUVUwfrYKGAIn' \
  -H 'X-Source: web' \
  -H 'X-User-ID: fca2925f-1eee-4654-9177-fece3fd6afc9' \
  -H 'X-msgid: 0892cc42-38c3-4214-8d29-c65a20d86191' \
  -H 'sec-ch-ua: "Google Chrome";v="113", "Chromium";v="113", "Not-A.Brand";v="24"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'ts: 2023-05-08T12:32:35+05:30' \
  --data-raw '{"request":{"content":{"name":"Copy of Text type QuestionSet","description":"texts","code":"4dee29fe-1e63-1c0b-f249-9459a232bfe2.copy","creator":"Tn content creator","createdFor":["01269878797503692810"],"createdBy":"fca2925f-1eee-4654-9177-fece3fd6afc9","organisation":["Tamil nadu"],"framework":"tn_k-12_5","mimeType":"application/vnd.sunbird.questionset"}}}' \
  --compressed`

Discussion Link -> https://github.com/orgs/Sunbird-Ed/discussions/465

@sajeshkayyath @krgauraw can you please help us to fix this . ( might be API onboarding script is missing in ED)
 
CC: @rajeevsathish @vinukumar-vs  @PoojaJManjooran @kameshbhr  @surendrasinghs   @rajnishdargan  @vaibhavbhuva @pallakartheekreddy

**Comment by krgauraw** [comment]
@princegupta1131 Can you please share the code changes PR to understand the changes? Also what is the final service end point which will be referred from http://localhost:3000/content/questionset/v1/copy/do_2133979147196088321344 ?

CC: @rajeevsathish @vinukumar-vs @PoojaJManjooran @kameshbhr @surendrasinghs @rajnishdargan @vaibhavbhuva @pallakartheekreddy @sajeshkayyath

**Reply by princegupta1131** [reply]
@krgauraw I have changed the service endpoint
from -> http://localhost:3000/content/content/v1/copy/do_2133979147196088321344' 
to     ->  http://localhost:3000/content/questionset/v1/copy/do_2133979147196088321344    
 & added a condition   if  `mimeType === 'application/vnd.sunbird.questionset'`.  
 CC: @rajeevsathish @vinukumar-vs @PoojaJManjooran @kameshbhr @surendrasinghs @rajnishdargan @vaibhavbhuva @pallakartheekreddy @sajeshkayyath

**Reply by krgauraw** [reply]
@princegupta1131 you should be using assessment-service end points for copying a questionset / question. You need to onboard question & questionset copy api's . Please find below reference

question copy: 
https://github.com/project-sunbird/sunbird-devops/blob/d3bae2018ebdc2c8bbfca8542d297a46dc2b1e5d/ansible/roles/kong-api/defaults/main.yml#L6322

questionset copy:
https://github.com/project-sunbird/sunbird-devops/blob/d3bae2018ebdc2c8bbfca8542d297a46dc2b1e5d/ansible/roles/kong-api/defaults/main.yml#L6552

Note: Please use above links for reference only. the script format may be different based on kong version.

Hope this helps.

CC: @rajeevsathish @vinukumar-vs @PoojaJManjooran @kameshbhr @surendrasinghs @rajnishdargan @vaibhavbhuva @pallakartheekreddy @sajeshkayyath

**Reply by princegupta1131** [reply]
@krgauraw Sure, let me try this thanks!

**Reply by princegupta1131** [reply]
Hi @krgauraw can you please share the link of release document for QuestionSet copy onboarding api

CC: CC: @rajeevsathish @vinukumar-vs @PoojaJManjooran @kameshbhr @surendrasinghs @rajnishdargan @vaibhavbhuva @pallakartheekreddy @sajeshkayyath @santhosh-tg

**Comment by vinukumar-vs** [comment]
> in SunbirdED -Portal on a copy of the Question set, getting 400 bad server error
> 
> As per the suggestion, we have done the code level changes to copy the question set.
> 
> **1. URL: {{host}}/questionset/v1/copy/{{questionset_id}} 2. Request Body: please use questionset instead of content**
> 
> **Post integration api throwing 401 Unauthorised Error**
> 
> **CURL**
> 
> `curl 'http://localhost:3000/content/questionset/v1/copy/do_2133979147196088321344' \ -H 'Accept: application/json' \ -H 'Accept-Language: en-US,en;q=0.9' \ -H 'Cache-Control: no-cache' \ -H 'Connection: keep-alive' \ -H 'Content-Type: application/json' \ -H 'Cookie: connect.sid=s%3AStewujMBObqoHrK7tIuqqI9m9bi3H49M.vsS4V%2F%2BDqhE5QyAqmaMFT9BmqecCG7zR81kzHORxMew' \ -H 'Origin: http://localhost:3000' \ -H 'Pragma: no-cache' \ -H 'Referer: http://localhost:3000/resources/play/questionset/do_2133979147196088321344' \ -H 'Sec-Fetch-Dest: empty' \ -H 'Sec-Fetch-Mode: cors' \ -H 'Sec-Fetch-Site: same-origin' \ -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36' \ -H 'X-App-Id: local.sunbird.portal' \ -H 'X-App-Version: 6.0.0' \ -H 'X-Channel-Id: 01269878797503692810' \ -H 'X-Device-ID: d22678ad8b5cc0f76f63dc0e7f4d1480' \ -H 'X-Org-code: 01269878797503692810' \ -H 'X-Request-ID: 0892cc42-38c3-4214-8d29-c65a20d86191' \ -H 'X-Session-ID: RB7pvdtVRopF24OkoGYEUVUwfrYKGAIn' \ -H 'X-Source: web' \ -H 'X-User-ID: fca2925f-1eee-4654-9177-fece3fd6afc9' \ -H 'X-msgid: 0892cc42-38c3-4214-8d29-c65a20d86191' \ -H 'sec-ch-ua: "Google Chrome";v="113", "Chromium";v="113", "Not-A.Brand";v="24"' \ -H 'sec-ch-ua-mobile: ?0' \ -H 'sec-ch-ua-platform: "macOS"' \ -H 'ts: 2023-05-08T12:32:35+05:30' \ --data-raw '{"request":{"content":{"name":"Copy of Text type QuestionSet","description":"texts","code":"4dee29fe-1e63-1c0b-f249-9459a232bfe2.copy","creator":"Tn content creator","createdFor":["01269878797503692810"],"createdBy":"fca2925f-1eee-4654-9177-fece3fd6afc9","organisation":["Tamil nadu"],"framework":"tn_k-12_5","mimeType":"application/vnd.sunbird.questionset"}}}' \ --compressed`
> 
> Discussion Link -> https://github.com/orgs/Sunbird-Ed/discussions/465
> 
> @sajeshkayyath @krgauraw can you please help us to fix this . ( might be API onboarding script is missing in ED)
> 
> CC: @rajeevsathish @vinukumar-vs @PoojaJManjooran @kameshbhr @surendrasinghs @rajnishdargan @vaibhavbhuva @pallakartheekreddy

Refer the below PR to onboard new API 
https://github.com/project-sunbird/sunbird-devops/pull/3764/files

Refer this documentation to on-board new API
https://project-sunbird.atlassian.net/wiki/spaces/DevOps/pages/1205207131/Best+Practces+For+Onboarding+API+s+In+Sunbird

**Reply by vinukumar-vs** [reply]
Is it similar issue?
https://github.com/orgs/Sunbird-inQuiry/discussions/24

**Reply by krgauraw** [reply]
@vinukumar-vs that was different issue.

**Comment by princegupta1131** [comment]
Hi @krgauraw unable to copy the question set in staging and was getting the below error
**"Invalid Props are : [publish_type, migrationVersion]"**  with help  of devops we added 

could you please update these 2 Props  [publish_type, migrationVersion] in the below variable and update the DevOps repo 

assessment.copy.props_to_remove=[“downloadUrl”, “artifactUrl”, “variants”,
“createdOn”, “collections”, “children”, “lastUpdatedOn”, “SYS_INTERNAL_LAST_UPDATED_ON”,
“versionKey”, “s3Key”, “status”, “pkgVersion”, “toc_url”, “mimeTypesCount”,
“contentTypesCount”, “leafNodesCount”, “childNodes”, “prevState”, “lastPublishedOn”,
“flagReasons”, “compatibilityLevel”, “size”, “publishChecklist”, “publishComment”,
“LastPublishedBy”, “rejectReasons”, “rejectComment”, “gradeLevel”, “subject”,
“medium”, “board”, “topic”, “purpose”, “subtopic”, “contentCredits”,
“owner”, “collaborators”, “creators”, “contributors”, “badgeAssertions”, “dialcodes”,
“concepts”, “keywords”, “reservedDialcodes”, “dialcodeRequired”, “leafNodes”, “sYS_INTERNAL_LAST_UPDATED_ON”, “prevStatus”, “lastPublishedBy”, “streamingUrl”]


@rajeevsathish @vinukumar-vs  @sajeshkayyath @santhosh-tg

**Reply by sajeshkayyath** [reply]
Path where this needs to be updated

https://github.com/project-sunbird/sunbird-devops/blob/release-6.0.0-inquiry/ansible/roles/stack-sunbird/templates/assessment-service_application.conf

**Reply by krgauraw** [reply]
@princegupta1131 this is taken in inQuiry release-6.0.0.
FYI https://github.com/project-sunbird/sunbird-devops/pull/3843
cc: @sajeshkayyath @PoojaJManjooran
